modules/imap/functions.php

```diff
@@ -1417,3 +1417,79 @@ function parse_sieve_config_host($host) {
     // $host = '$scheme://'.$host;
     return [$host, $port, $tls];
 }}
+
+if (!hm_exists('connect_to_imap_serve')) {
+    function connect_to_imap_server($address, $name, $port, $user, $pass, $tls, $imap_sieve_host, $enableSieve, $context, $errno = null, $errstr = null) {
+        if ($con = fsockopen($address, $port, $errno, $errstr, 5)) {
+              $imap_list = array(
+                  'name' => $name,
+                  'server' => $address,
+                  'hide' => false,
+                  'port' => $port,
+                  'user' => $user,
+                  'pass' => $pass,
+                  'tls' => $tls);
+                  
+              Hm_IMAP_List::add($imap_list);
+              $servers = Hm_IMAP_List::dump(false, true);
+              $ids = array_keys($servers);
+              $imap_server_id = array_pop($ids);
+              
+             return $imap_server_id;
+        }
+    }
+}
+
+
+if (!hm_exists('connect_to_jmap_serve')) {
+    function connect_to_jmap_server($jmap_address, $hide_from_c_page, $name, $user, $pass) {
+       $hidden = false;
+                       
+       if($hide_from_c_page) {
+           $hidden = true;
+       }
+       
+       $parsed = parse_url($jmap_address);
+       
+       if (array_key_exists('host', $parsed) && @get_headers($jmap_address)) {
+           Hm_IMAP_List::add(array(
+               'name' => $name,
+               'server' => $jmap_address,
+               'hide' => $hidden,
+               'type' => 'jmap',
+               'port' => false,
+               'tls' => false));
+           $servers = Hm_IMAP_List::dump(false, true);
+           $ids = array_keys($servers);
+           $jmap_server_id = array_pop($ids);
+           
+           return $jmap_server_id;
+       }
+    }
+}
+
+if (!hm_exists('authenticate_to_imap_server')) {
+    function authenticate_to_imap_server($user, $pass, $imap_server_id, $context) {
+        if (in_server_list('Hm_IMAP_List', $imap_server_id, $user)) {
+              Hm_Msgs::add('ERRThis server and username are already configured');
+              return;
+        } else{
+            Hm_IMAP_List::clean_up();
+            $cache = Hm_IMAP_List::get_cache($context->cache, $imap_server_id);
+            $imap = Hm_IMAP_List::connect($imap_server_id, $cache, $user, $pass, true);
+            
+            if (imap_authed($imap)) {
+                  $servers = Hm_IMAP_List::dump(false, true);
+                  $context->user_config->set('imap_servers', $servers);
```

Comment:
Repository pattern was introduced in master and we no longer save the servers externally - please check other places how the storage of server list is done.